### ISSUE TITLE
What's the plan for 2.0 and future development?

### ISSUE THREAD

**Issue by flight16** [maebert/jrnl#339]
From what I gather, 2.0 includes breaking changes.  The 2.0 rc1 hasn't seen any commits in several months.  There are several big changes like importing and journals in multiple files that are planned but haven't had any updates in 1/2 year.

I was just wondering if @maebert could share his plans for 2.0 and the future development of jrnl.

I ask because I want to work on improving the storage layer to gain the option of storing journals across multiple files #170, but I'm not sure if I should develop against master or 2.0 branch.

**Comment by maebert** [comment]
Hi there!

First of all, I'm sorry for being a little unresponsive and not giving jrnl as much time as it deserves these days. I got busy [starting a company](knot63.com) and [traversing the world](http://1450.me/#map).

Here is the master plan for 2.0:

1. Switch the encryption engine from PyCrypto to [cryptograph](https://cryptography.io/en/latest). Why? PyCrypto is a pain to install on different platforms. Cryptography uses OpenSSL as a backend and is much easier to install. __This step is done on the 2.0 branch.__
2. Have a generalised back-end interface so that writing a back-end that has multiple files is nothing but writing a new class with a different load / save method. __This is basically also done.__ I'd like to make it a little smarter so that it doesn't have to load _all_ files when you're just saving one for example.
3. Make YAML the new config format. It's more robust and easier to understand. __Done, too.__
4. Writing upgrade methods for converting journals encrypted with 1.x to the new format. __This is not done yet and is the reasons tests are failing right now. Help appreciated__
5. Write more tests. Help always appreciated.

I know there are a lot of PRs and a lot of people who'd like to support jrnl, and I'm super grateful for that. I'm going to manage as many of the PRs as possible today and then follow up here again!

**Comment by flight16** [comment]
You sound busy indeed.  Thanks for finding the time to reply.

That clears a lot of things up for me.  I think the only ones I might possibly be able to help with are 4 and 5.  Other than that I'd like to write another journal format or two and fix a few outstanding issues if I can find the time.

I couldn't find mention of YAML in 2.0-rc1 (I assume that is what you mean by the 2.0 branch).  Maybe it's not pushed yet?

**Comment by maebert** [comment]
It's in `util.load_config`:

```python
def load_config(config_path):
    with open(config_path) as f:
        config = yaml.load(f)
    return config
```

Easy as that ;)

**Comment by flight16** [comment]
So it sounds like writing code to upgrade encrypted journals is the last item left?  I would think upgrading date format, too.  If so, I'll see if I can find time this week to work on that.

**Comment by flight16** [comment]
@maebert It would be cool if the [2.0.0 milestone tag](https://github.com/maebert/jrnl/milestones/2.0.0) was up-to-date so I knew where to focus my energy. :)

**Comment by maebert** [comment]
Writing code for upgrading encrypted journals is now done, too! And upgrading the date format, too. All there. All done.

What's really left to do is TESTING. Lots and lots of testing. And then more testing! Closing this issue now though :)

**Comment by maebert** [comment]
@flight16 Closed some issues that were fixed in 2.0!

**Comment by flight16** [comment]
Thank you for the replies and closing some issues!

Sorry for being persistent about this, but...

I still see 6 issues in the [milestone for 2.0.0](https://github.com/maebert/jrnl/milestones/2.0.0) that span bug, feature, large and small.  I assume that is the "things to resolve before 2.0.0 can be released" list, so I want to verify that the list is up-to-date and you expect all of them to be implemented before a release can happen.

If the focus is on getting 2.0.0 wrapped up and priority is given to breaking changes, then I would think the current 2.0.0 issues should be split out as follows:

### v2.0.0

* #297 No default configuration for 2.0rc1 bug (show-stopping bug, although this seemed to be implemented when I tryed 2.0rc1 branch)
* #282 Plugin Framework Integration (big changes.  I saw some plugin-like hooks and code in 2.0rc1 branch.  Not sure what else there is left to do here)

### v2.1.0 and beyond

* #257 Import? Integration
* #171 -quiet option Ideas Integration
* #37 entries containing data Integration
* #170 Use several files instead of only one? (can be supported with a new journal type, ideally minimal changes to core code will be needed)

I've been using jrnl daily for the last two weeks and like it a lot (thanks!).  I have many more bug fixes and features I'd like to work on, but I'd first like to help 2.0.0 get released so there is a single branch to develop against. : )

**Comment by maebert** [comment]
Totally agree :) I'd like to get #257 into 2.0 though. It's not a difficult feature to implement and it proves that the current framework makes sense. The rest: I agree. I'll change the milestones.
